gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,216 @@
+  private final RequestT prototype;
+  private Batch<ElementT, ElementResultT, RequestT> currentOpenBatch;
+
+  private final AtomicInteger numOfOutstandingBatches = new AtomicInteger(0);
+  private final Object flushLock = new Object();
+  private boolean isClosed = false;
+
+  private BatcherImpl(Builder<ElementT, ElementResultT, RequestT, ResponseT> builder) {
+    this.prototype = checkNotNull(builder.prototype, "RequestPrototype cannot be null.");
+    this.callable = checkNotNull(builder.unaryCallable, "UnaryCallable cannot be null.");
+    this.batchingDescriptor =
+        checkNotNull(builder.batchingDescriptor, "BatchingDescriptor cannot be null.");
+  }
+
+  /** Builder for a BatcherImpl. */
+  public static class Builder<ElementT, ElementResultT, RequestT, ResponseT> {
+    private BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor;
+    private UnaryCallable<RequestT, ResponseT> unaryCallable;
+    private RequestT prototype;
+
+    private Builder() {}
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setBatchingDescriptor(
+        BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor) {
+      this.batchingDescriptor = batchingDescriptor;
+      return this;
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setUnaryCallable(
+        UnaryCallable<RequestT, ResponseT> unaryCallable) {
+      this.unaryCallable = unaryCallable;
+      return this;
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setPrototype(RequestT prototype) {
+      this.prototype = prototype;
+      return this;
+    }
+
+    public BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT> build() {
+      return new BatcherImpl<>(this);
+    }
+  }
+
+  public static <EntryT, EntryResultT, RequestT, ResponseT>
+      Builder<EntryT, EntryResultT, RequestT, ResponseT> newBuilder() {
+    return new Builder<>();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ElementResultT> add(ElementT element) {
+    Preconditions.checkState(!isClosed, "Cannot add elements on a closed batcher.");
+
+    if (currentOpenBatch == null) {
+      currentOpenBatch = new Batch<>(batchingDescriptor.newRequestBuilder(prototype));
+    }
+
+    SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
+    currentOpenBatch.add(element, result);
+    return result;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void flush() throws InterruptedException {
+    sendBatch();
+    awaitAllOutstandingBatches();
+  }
+
+  /** Sends accumulated elements asynchronously for batching. */
+  private void sendBatch() {
+    if (currentOpenBatch == null) {
+      return;
+    }
+    final Batch<ElementT, ElementResultT, RequestT> accumulatedBatch = currentOpenBatch;
+    currentOpenBatch = null;
+    numOfOutstandingBatches.incrementAndGet();
+
+    final ApiFuture<ResponseT> batchResponse =
```

Comment:
If this throws an exception, `numOfOutstandingBatches` incremented above (L138) will never be decremented, is it right? Can we do something about it?

Reply:
Thanks for pointing it out, Now incrementing the counter only after receiving `ApiFuture<ResponseT>`.

Reply:
Wait, I think it can be even worse now, especially taking into account that `onBatchCompletion()` checks for 0 equality:
```java
if (numOfOutstandingBatches.decrementAndGet() == 0)
```
not
```java
if (numOfOutstandingBatches.decrementAndGet() <= 0)
```
I believe in case if that thing executes really fast, the callback can be called before the caller receives the `batchResponse` future and manages to increment the batches, so `numOfOutstandingBatches.decrementAndGet()` will return -1.

In general i had in mind some finally block or something, but the changed version looks even more dangerous than original (the possible exception was very theoretical thing, but now it is a real danger of getting -1, when 0 is expected).

Reply:
The order of operations is
1. increment numOfOutstandingBatches
2. add the callback
3. at some point in the future, most likely on a different thread, invoke onBatchCompletion which calls numOfOutstandingBatches.decrementAndGet

So it doesn't matter how fast things execute, there is no race. It's possible for the callback to to be invoked before the caller gets the batchResponse future, but it doesn't matter. The count will be 0.

Please let me know what I'm missing because I dont see a problem here

Reply:
@igorbernstein2/ @vam-google I followed Igor's last comment and I believe using a public Xor should be fine. Please have a fresh look.